File: Java/1329.SortMatrixDiagonally/Solution.java
/* Sort: Time:O(n^2logn), Space:O(n), where n is max(mat.length, mat[0].length)
 * 1. Retrieve each diagnoals
 * 2. Sort them and reassign them to mat
 */     

import java.util.*; // Stack

public class Solution {
    public int[][] diagonalSort(int[][] mat) {
        for(int y = mat.length - 1; y >= 0; --y){
            int length = Math.min(mat.length - y,  mat[0].length); 
            List<Integer> diagnoal = new ArrayList<>();
            for(int x = 0; x < length; ++x){
                diagnoal.add(mat[y + x][x]);
            }
            Collections.sort(diagnoal);
            for(int x = 0; x < length; ++x){
                mat[y + x][x] = diagnoal.get(x);
            }
        }
        
        for(int x = 1; x < mat[0].length; ++x){
            int length = Math.min(mat[0].length - x, mat.length);
            List<Integer> diagnoal = new ArrayList<>();
            for(int y = 0; y < length; ++y){
                diagnoal.add(mat[y][x + y]);
            }
            Collections.sort(diagnoal);
            for(int y = 0; y < length; ++y){
                mat[y][x + y] = diagnoal.get(y);
            }
        }
        return mat;
    }

    public static void main(String[] args){
        Solution sol = new Solution();
        int[][] mat = {{3, 3, 1, 1}, {2, 2, 1, 2}, {1, 1, 1, 2}};
        System.out.println("before:");
        for(int[] row: mat){
            System.out.println(Arrays.toString(row));
        }
        mat = sol.diagonalSort(mat);
        System.out.println("after:");
        for(int[] row: mat){
            System.out.println(Arrays.toString(row));
        }
    }
}

File: Go/1329.SortMatrixDiagonally/Solution.go
/* Sort: Time:O(n^2logn), Space:O(n), where n is max(mat.length, mat[0].length)
 * 1. Retrieve each diagnoals
 * 2. Sort them and reassign them to mat
 */

package main

import (
    "fmt"
    "sort"
)

func min(x int, y int) int {
    if(x > y){
        return y
    }else{
        return x
    }
}

func diagonalSort(mat [][]int) [][]int {
    for y := len(mat) - 1; y >= 0; y-- {
        length, diagnoal := min(len(mat) - y, len(mat[0])), make([]int, 0, 0)
        for x := 0; x < length; x++ {
            diagnoal = append(diagnoal, mat[y + x][x])
        }
        sort.Ints(diagnoal)
        for x := 0; x < length; x++ {
            mat[y + x][x] = diagnoal[x]
        }
    }

    for x := 1; x < len(mat[0]); x++ {
        length, diagnoal := min(len(mat[0]) - x, len(mat)), make([]int, 0, 0)
        for y := 0; y < length; y++ {
            diagnoal = append(diagnoal, mat[y][x + y])
        }
        sort.Ints(diagnoal)
        for y := 0; y < length; y++ {
            mat[y][x + y] = diagnoal[y]
        }
    }
    return mat
}

func main() {
    mat := [][]int{{3, 3, 1, 1}, {2, 2, 1, 2}, {1, 1, 1, 2}}
    fmt.Println("before:", mat)
    fmt.Println("after:", diagonalSort(mat))
}
